tensorflow/compiler/mlir/xla/transforms/input_inline_fusion_pass.cc

```diff
@@ -0,0 +1,268 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements loop fusion.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Analysis/AffineAnalysis.h"
+#include "mlir/Analysis/AffineStructures.h"
+#include "mlir/Analysis/LoopAnalysis.h"
+#include "mlir/Analysis/Utils.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/AffineExpr.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/LoopFusionUtils.h"
+#include "mlir/Transforms/LoopUtils.h"
+#include "mlir/Transforms/Passes.h"
+#include "mlir/Transforms/Utils.h"
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"  // TF:llvm-project
+#include "mlir/Dialect/LoopOps/EDSC/Builders.h" 
+#include "mlir/IR/Attributes.h"  // TF:llvm-project
+#include "mlir/IR/Location.h"  // TF:llvm-project
+#include "mlir/IR/MLIRContext.h"  // TF:llvm-project
+#include "mlir/IR/PatternMatch.h"  // TF:llvm-project
+#include "mlir/IR/StandardTypes.h"  // TF:llvm-project
+#include "mlir/Pass/Pass.h"  // TF:llvm-project
+#include "tensorflow/compiler/mlir/xla/ir/lhlo_ops.h"
+#include "tensorflow/compiler/mlir/xla/transforms/map_xla_to_scalar_op.h"
+#include <iostream>
+
+namespace mlir {
+namespace xla_lhlo {
+
+namespace {
+
+struct InputInlineFusion : public FunctionPass<InputInlineFusion> {
+  void runOnFunction() override;
+};
+
+} // end anonymous namespace
+
+std::unique_ptr<OpPassBase<FuncOp>> createInputInlineFusionPass() {
+  return std::make_unique<InputInlineFusion>();
+}
+
+namespace {
+
+class InputInlineFusionPattern: public RewritePattern {
+ public:
+  explicit InputInlineFusionPattern(MLIRContext *context)
+      : RewritePattern(loop::ForOp::getOperationName(), 1, context) {}
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    // skip if not the most outter ForOp
+    auto parent = op->getParentOp();
+    if (!isa<FuncOp>(parent)) {
+      return failure();
+    }
+
+
+    SmallVector<mlir::LoadOp, 4> load_ops;
+    op->walk([&](mlir::LoadOp load_op) {
+      load_ops.push_back(load_op);
+    });
+
+    for (auto load_op : load_ops) {
+      auto lhlo_op = getFusibleOperation(load_op);
+      if (lhlo_op != nullptr) {
+        // 1, in case of:
+        //      A = ...
+        //      B = op(A)
+        //      C = op(A, B)
+        //    C should fuse B first before fusing A.
+        //    This is the same logic as in instruction_fusion pass of XLA
+        //
+        // 2, When multiple loads consumes the same result of lhlo_op and 
+        //    the load indices are also identical, the ir should be
+        //    emitted only once. Other LoadOps should used cached Value.
+
+        // 'load_ops' that can consume the same cached value
+        std::vector<Operation*> load_ops;
```

Comment:
Choose another name to avoid potential bug, since "load_ops" is already used before.